Unicode/ustring.c

```diff
@@ -655,18 +655,19 @@ char *u2utf8_copy(const unichar_t *ubuf) {
 
 char *u2utf8_copyn(const unichar_t *ubuf,int len) {
 /* Make a utf8 string copy of unichar string ubuf[0..len] */
-    char *utf8buf, *pt;
+    char *utf8buf, *pt, *pt2;
 
     if ( ubuf==NULL || len<=0 || (utf8buf=pt=(char *)malloc(len*6+1))==NULL )
 	return( NULL );
 
-    while ( (pt=utf8_idpb(pt,*ubuf++,0)) && --len );
-    if ( pt ) {
-	*pt = '\0';
-	return( utf8buf );
-    }
-    free( utf8buf );
-    return( NULL );
+    while ( (pt2=utf8_idpb(pt,*ubuf++,0)) && --len )
+	pt = pt2;
+
+    if ( pt2 )
+	pt = pt2;
+
```

Comment:
Now that I'm on my PC I can confirm that I did not cause this. I thought I may have in #4160, but that didn't touch this function.

I think a developer `TRACE` warning should be added here on truncation. It seems like the kind of thing to make debugging difficult when we're expecting a string of one length and getting a length of another. We may even want to consider changing the function signature to int *len so lengths can be verified as needed.